### DISCUSSION TITLE
Can I use third party Xamarin.Android UI componets (not for Xamarin.Forms) in AvaloniaUI?

### DISCUSSION THREAD

**Discussion by ivoryguard** [AvaloniaUI/Avalonia#10363] [Q&A] [answered]
Hello.

I am considering AvaloniaUI platform to develop an Android Tablet App.

It will be currently for Android only but it can be extended to other platforms in future.
So, I need to have a cross-platform option to support MacOS, iOS and Windows.

I am very familiar with WPF Xaml, I prefer Xaml to Android UI system of Xamarin.Android.

However, I am afraid that There are not enough 3rd party components for AvaloniaUI.

In detail, I need a full-featured PDF viewer/renderer component, but I cannot find anything at AvaloniaUI components and other 3rd party components for AvaloniaUI.

But, there is a few PDF viewer/render component for Xamarin.Android and Android Native.

So, If I can use it through a renderer like Xamarin.Forms, I will get much more component options in available at AvaloniaUI.

Could you please let me know whether it is possible to add Xamarin.Android UI components or Android Native View in Xaml Visual Tree with less performance loss?

Best regards.

**Comment by kekekeks** [comment, marked as answer]
Embedding native controls is supported. You can embed any component that provides `UIView`/`android.view.View`.

**Reply by ivoryguard** [reply]
Thank you very much!
Then, is it possible to exchange data between XAML code behind and UIView/Android.view.View?
(e.g. event handler, commanding, messaging subscription like MediatR and so on)

**Reply by maxkatz6** [reply]
Yes, it's not avalonia part of the problem. You need to manage it in c# code.
